tests/TestKdbx3XmlReader.cpp

```diff
@@ -0,0 +1,22 @@
+/*
+ *  Copyright (C) 2017 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <QTest>
+
+#include "TestKeePass2XmlReader.h"
+
+QTEST_GUILESS_MAIN(TestKdbx3XmlReader)
```

Comment:
Where did the rest of these tests go?

Reply:
Look at CMakeLists.

Reply:
OK... so there are two cpp files, one is basically blank. Why are the TestKdbx3XmlReader member functions defined in TestKeePass2XmlReader.cpp??

Reply:
I guess you didn't take a look at the 4 commits that were left out of this PR.

`TestKeePass2XmlReader` has tests that instantiate `KeePass2XmlReader` directly like [this one](https://github.com/angelsl/keepassxc/blob/master/tests/TestKeePass2XmlReader.cpp#L415). In order to avoid having to duplicate the test functions between `Kdbx3XmlReader` and `Kdbx4XmlReader`, I pulled out the XML read and writes into [virtual functions](https://github.com/angelsl/keepassxc/blob/kdbx42/tests/TestKeePass2XmlReader.h#L50) that [use the correct `XmlReader` class](https://github.com/angelsl/keepassxc/blob/kdbx42/tests/TestKeePass2XmlReader.cpp#L81). I decided to keep them in the same source file so that if these functions are ever changed, it is less likely that one half is forgotten, and so that it is easier to verify that they are identical aside from the different `XmlReader` class used.

The better solution, arguably, would have been to use templates, but Qt `moc` does not support templates.

Another solution would have been making the two `XmlReader` classes a subclass of a base `XmlReader` class like the other `Reader` classes, but since the `XmlReader`s are only ever used by the `Reader` classes themselves, I didn't think that was appropriate: the `XmlReader` classes are implementation details of the `Reader` classes and it does not make sense for them to be sibling classes. The only way in which they are similar is that they both take XML data (that *happens* to be similarly structured) and spit out a database. Arguably the `XmlReader` classes should be nested classes of the `Reader` classes, but I digress.

Finally, in case you are wondering why I didn't stick both `QTEST_GUILESS_MAIN` in the `TestKeePass2XmlReader` file, `QTEST_GUILESS_MAIN` is a macro that expands to a definition of `int main`.